File: app/src/main/java/com/devusercode/upchat/ConversationActivity.java
package com.devusercode.upchat;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.devusercode.upchat.adapter.MessageAdapter;
import com.devusercode.upchat.adapter.WrapLayoutManager;
import com.devusercode.upchat.models.Message;
import com.devusercode.upchat.models.User;
import com.devusercode.upchat.utils.ConversationUtil;
import com.devusercode.upchat.utils.UserUtils;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class ConversationActivity extends AppCompatActivity {
    private final String TAG = this.getClass().getSimpleName();

    private final FirebaseAuth auth = FirebaseAuth.getInstance();
    private final FirebaseDatabase firebaseDatabase = FirebaseDatabase.getInstance();
    private final DatabaseReference conversations = firebaseDatabase.getReference("conversations");

    /* Toolbar */
    private ImageView back_button;
    private TextView participant_name;
    private ImageView profile_image;

    /* Content */
    private MessageAdapter adapter;
    private RecyclerView recyclerview;
    private Button attach_button;
    private TextInputEditText message_input;
    private Button send_button;

    private final Intent intent = new Intent();

    private User participant;
    private User user;
    private String currentConversationId;
    private boolean chatExists = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        FirebaseApp.initializeApp(this);

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_conversation);

        // FirebaseDatabase.getInstance().setPersistenceEnabled(true);

        // intent for data
        Intent data_intent = getIntent();

        // get uid and get participant user
        if (data_intent != null && data_intent.getExtras() != null) {
            String uid = data_intent.getStringExtra("uid");
            getParticipant(uid);
        }
    }

    @Override
    public void startActivity(Intent intent) {
        super.startActivity(intent);
        overridePendingTransition(R.anim.right_in, R.anim.left_out);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.left_in, R.anim.right_out);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

    private void initialize() {
        CoordinatorLayout coordinator = findViewById(R.id._coordinator);
        AppBarLayout app_bar = findViewById(R.id.app_bar);
        Toolbar toolbar = findViewById(R.id.toolbar);

        setSupportActionBar(toolbar);

        back_button = toolbar.findViewById(R.id.back_button);
        participant_name = toolbar.findViewById(R.id.participant_name);
        profile_image = toolbar.findViewById(R.id.profile_image);

        attach_button = findViewById(R.id.attach_button);
        message_input = findViewById(R.id.message_input);
        send_button = findViewById(R.id.send_button);
        recyclerview = findViewById(R.id.chat_recyclerview);

        LinearLayoutManager layoutManager = new WrapLayoutManager(getApplicationContext(), LinearLayoutManager.VERTICAL, false);

        recyclerview.setLayoutManager(layoutManager);
        recyclerview.setItemAnimator(new DefaultItemAnimator());

        recyclerview.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
            if (bottom < oldBottom) {
                recyclerview.scrollBy(0, oldBottom - bottom);
            }
        });

        back_button.setOnClickListener(view -> {
            super.onBackPressed();
        });

        send_button.setOnClickListener(view -> {
            if (!chatExists) {
                currentConversationId = ConversationUtil.newConversation(user, participant);
                Log.d(TAG, "Conversation (" + currentConversationId + ") created");
                loadConversation(currentConversationId);
            }

            if (!message_input.getText().toString().isEmpty()) {
                sendMessage(message_input.getText().toString());
                message_input.setText("");
            }
        });

        participant_name.setText(participant.getUsername());

        if (!participant.getPhotoUrl().isEmpty()) {
            Glide.with(getApplicationContext())
                    .load(Uri.parse(participant.getPhotoUrl()))
                    .placeholder(R.drawable.ic_account_circle_black)
                    .circleCrop()
                    .into(profile_image);
        } else {
            profile_image.setImageResource(R.drawable.ic_account_circle_black);
        }

        chatExists = ConversationUtil.conversationExistsForBoth(user, participant);

        if (chatExists) {
            currentConversationId = ConversationUtil.getConversationId(participant, user.getConversations());
            loadConversation(currentConversationId);
        }
    }

    private void loadConversation(String cid) {
        Query messages = conversations
                .child(currentConversationId)
                .child(Key.Conversation.MESSAGES);

        FirebaseRecyclerOptions<Message> options = new FirebaseRecyclerOptions.Builder<Message>()
                .setQuery(messages, Message.class)
                .build();

        adapter = new MessageAdapter(getApplicationContext(), options);
        adapter.setConversationId(cid);
        adapter.setParticipant(participant);
        adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
            @Override
            public void onItemRangeInserted(int positionStart, int itemCount) {
                super.onItemRangeInserted(positionStart, itemCount);
                recyclerview.smoothScrollToPosition(positionStart);
            }
        });

        recyclerview.setAdapter(adapter);
        adapter.startListening();
    }

    private void sendMessage(String message) {
        Map<String, String> data = new HashMap<>();

        DatabaseReference messagesRef = FirebaseDatabase.getInstance().getReference()
                .child(Key.Conversation.CONVERSATIONS)
                .child(currentConversationId)
                .child(Key.Conversation.MESSAGES);

        String messageId = messagesRef.push().getKey();

        if (messageId != null && messageId.startsWith("-")) {
            messageId = messageId.substring(1);
        }

        // Save the message data
        data.put(Key.Message.MESSAGE, message.trim());
        data.put(Key.Message.ID, messageId);
        data.put(Key.Message.TYPE, "text");
        data.put(Key.Message.SENDER_ID, user.getUid());
        data.put(Key.Message.TIMESTAMP, String.valueOf(System.currentTimeMillis()));

        messagesRef.child(messageId).setValue(data);

        try {
            sendNotification(message);
        } catch (Exception e) {
            Log.e(TAG, e.getMessage());
        }
    }

    private void sendNotification(String message) throws JSONException {
        JSONObject jsonObject = new JSONObject();

        JSONObject notificationObj = new JSONObject();
        notificationObj.put("title", user.getUsername());
        notificationObj.put("body", message);

        JSONObject dataObj = new JSONObject();
        dataObj.put("userId", user.getUid());

        jsonObject.put("notification", notificationObj);
        jsonObject.put("data", dataObj);
        jsonObject.put("to", participant.getDeviceToken());

        MediaType JSON = MediaType.get("application/json; charset=utf-8");
        OkHttpClient client = new OkHttpClient();
        String url = "https://fcm.googleapis.com/fcm/send";
        RequestBody body = RequestBody.create(jsonObject.toString(), JSON);
        Request request = new Request.Builder()
                .url(url)
                .post(body)
                .header("Authorization", "Bearer AIzaSyBhEaPNNyzZjZryMITa9yhhhALfUoBUWwg")
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {

            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {

            }
        });
    }

    private void getParticipant(String uid) {
        UserUtils.getUserByUid(uid, result -> {
            if (result.getUser() != null) {
                participant = result.getUser();
                getUser();
            } else {
                Log.e(TAG, result.getError().getMessage());
            }
        });
    }

    private void getUser() {
        UserUtils.getUserByUid(auth.getCurrentUser().getUid(), result -> {
            if (result.getUser() != null) {
                user = result.getUser();
                initialize();
            } else {
                user = null;
                Log.e(TAG, result.getError().getMessage());
            }
        });
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (adapter != null) {
            adapter.startListening();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        adapter.stopListening();
    }

    @Override
    protected void onPause() {
        super.onPause();
        // Remove the following line to prevent stopping adapter listening in onPause()
        // adapter.stopListening();
    }

    @SuppressLint("NotifyDataSetChanged")
    @Override
    protected void onResume() {
        super.onResume();
        if (adapter != null) {
            adapter.startListening();
            adapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/devusercode/upchat/ConversationActivity.kt
package com.devusercode.upchat

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewTreeObserver
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.size
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.AdapterDataObserver
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.devusercode.upchat.adapter.MessageAdapter
import com.devusercode.upchat.adapter.WrapLayoutManager
import com.devusercode.upchat.models.Message
import com.devusercode.upchat.models.User
import com.devusercode.upchat.security.ConversationKeyManager
import com.devusercode.upchat.utils.ConversationUtil
import com.devusercode.upchat.utils.ErrorCodes
import com.devusercode.upchat.utils.StorageController
import com.devusercode.upchat.utils.UserUtils
import com.devusercode.upchat.utils.Util
import com.devusercode.upchat.utils.setComposeContent
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.textfield.TextInputEditText
import com.google.firebase.FirebaseApp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.Query
import com.google.firebase.database.ValueEventListener
import androidx.core.net.toUri

@RequiresApi(Build.VERSION_CODES.O)
class ConversationActivity : AppCompatActivity() {
    private val TAG = this.javaClass.simpleName
    private val auth = FirebaseAuth.getInstance()
    private val firebaseDatabase = FirebaseDatabase.getInstance()
    private val conversations = firebaseDatabase.getReference("conversations")
    private val users = firebaseDatabase.getReference("users")

    /* Toolbar */
    private lateinit var backButton: Button
    private lateinit var participantName: TextView
    private lateinit var profileImage: ImageView
    private lateinit var statusOnline: TextView

    /* Content */
    private var adapter: MessageAdapter? = null
    private lateinit var recyclerview: RecyclerView
    private lateinit var attachButton: Button
    private lateinit var messageInput: TextInputEditText
    private lateinit var sendButton: Button

    private var participant: User? = null
    private var user: User? = null

    private lateinit var contentView: View
    private lateinit var storageController: StorageController
    private lateinit var conversationUtil: ConversationUtil
    private var conversationSecret: String? = null

    private var file: Uri? = null
    private var filePickerLauncher: ActivityResultLauncher<String> =
        registerForActivityResult(ActivityResultContracts.GetContent()) { result ->
            Util.showMessage(this, "Image selected...")
            file = result
        }

    private lateinit var currentConversationId: String
    private var chatExists = false

    override fun onCreate(savedInstanceState: Bundle?) {
        FirebaseApp.initializeApp(this)

        super.onCreate(savedInstanceState)
        contentView = setComposeContent(R.layout.activity_conversation)

        storageController = StorageController.getInstance(this)!!

        // get uid and get participant user
        if (intent != null && intent.extras != null) {
            val uid = intent.getStringExtra("uid")!!

            UserUtils.getUserByUid(uid) { result ->
                if (result.code == ErrorCodes.SUCCESS) {
                    participant = result.user

                    if (storageController.contains("user")) {
                        user = storageController.getUser("user")
                        Log.d(TAG, "User from Storage" + user?.info.toString())

                        initialize()
                    } else {

                        UserUtils.getUserByUid(auth.currentUser!!.uid) { result2 ->
                            if (result2.code == ErrorCodes.SUCCESS) {
                                user = result2.user
                                initialize()
                            } else {
                                user = null
                                Log.e(TAG, result2.error?.message!!)
                            }
                        }
                    }
                } else {
                    Log.e(TAG, result.error?.message!!)
                }
            }
        } else {
            onBackPressedDispatcher.onBackPressed()
        }
    }

    override fun startActivity(intent: Intent) {
        super.startActivity(intent)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            overrideActivityTransition(OVERRIDE_TRANSITION_OPEN, R.anim.right_in, R.anim.left_out)
        } else {
            @Suppress("DEPRECATION")
            overridePendingTransition(R.anim.right_in, R.anim.left_out)
        }
    }

    override fun finish() {
        super.finish()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            overrideActivityTransition(OVERRIDE_TRANSITION_OPEN, R.anim.left_in, R.anim.right_out)
        } else {
            @Suppress("DEPRECATION")
            overridePendingTransition(R.anim.left_in, R.anim.right_out)
        }
    }

    private fun initialize() {
        // val coordinator = findViewById<CoordinatorLayout>(R.id._coordinator)
        // val appBar = findViewById<AppBarLayout>(R.id.app_bar)
        val toolbar = contentView.findViewById<Toolbar>(R.id.toolbar)

        setSupportActionBar(toolbar)

        toolbar.setOnClickListener {
            val intent = Intent(this, UserProfileActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.putExtra("uid", participant?.uid)
            startActivity(intent)
        }

        backButton = toolbar.findViewById(R.id.back_button)
        participantName = toolbar.findViewById(R.id.participant_name)
        profileImage = toolbar.findViewById(R.id.profile_image)
        statusOnline = toolbar.findViewById(R.id.status_online)

        attachButton = contentView.findViewById(R.id.attach_button)
        messageInput = contentView.findViewById(R.id.message_input)
        sendButton = contentView.findViewById(R.id.send_button)
        recyclerview = contentView.findViewById(R.id.chat_recyclerview)

        attachButton.setOnClickListener { filePickerLauncher.launch("image/*") }

        val layoutManager: LinearLayoutManager =
            WrapLayoutManager(applicationContext, LinearLayoutManager.VERTICAL, false)

        recyclerview.layoutManager = layoutManager
        recyclerview.itemAnimator = DefaultItemAnimator()

        // recyclerview.addOnScrollListener(object : RecyclerView.OnScrollListener() {})
        // view: View?, left: Int, top: Int, right: Int, bottom: Int, oldLeft: Int, oldTop: Int, oldRight: Int, oldBottom: Int
        recyclerview.addOnLayoutChangeListener { _: View?, _: Int, _: Int, _: Int, bottom: Int, _: Int, _: Int, _: Int, oldBottom: Int ->
            if (bottom < oldBottom) {
                recyclerview.scrollBy(0, oldBottom - bottom)
            }
            if (bottom > oldBottom)
                recyclerview.scrollBy(0, bottom - oldBottom)
        }

        // recyclerview.smoothScrollToPosition(recyclerview.adapter!!.itemCount - 1);

        backButton.setOnClickListener { onBackPressedDispatcher.onBackPressed() }

        sendButton.setOnClickListener {
            if (!chatExists) {
                chatExists = true

                ConversationUtil.newConversation(user!!, participant!!)
                    .addOnSuccessListener { conversationId ->
                        currentConversationId = conversationId
                        Log.d(TAG, "Conversation ($currentConversationId) created")
                        prepareConversation(currentConversationId)
                        Util.showMessage(this, "Setting up secure channel...")
                    }
                    .addOnFailureListener { error ->
                        chatExists = false
                        Log.e(TAG, "Error creating conversation", error)
                        Util.showMessage(
                            this,
                            error.message ?: getString(R.string.conversation__error_generic)
                        )
                    }
                return@setOnClickListener
            }

            if (!::conversationUtil.isInitialized) {
                prepareConversation(currentConversationId)
                Util.showMessage(this, "Secure channel is initializing. Please wait...")
                return@setOnClickListener
            }

            val message: String = messageInput.text.toString()

            if (file != null) {
                if (message.isNotEmpty()) {
                    conversationUtil.sendImage(file!!, message)
                    messageInput.setText("")
                } else {
                    conversationUtil.sendImage(file!!, null)
                }
                file = null
                return@setOnClickListener
            }

            if (message.isNotEmpty()) {
                conversationUtil.sendMessage(message)
                    .addOnSuccessListener { messageInput.setText("") }
                    .addOnFailureListener { error ->
                        Log.e(TAG, "Failed to send message", error)
                        Util.showMessage(
                            this,
                            error.message ?: getString(R.string.conversation__error_generic)
                        )
                    }
            }
        }

        participantName.text = participant!!.username

        if (participant?.photoUrl!!.isNotEmpty()) {
            Glide.with(applicationContext)
                .load(participant!!.photoUrl!!.toUri())
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .placeholder(R.drawable.ic_account_circle_white)
                .circleCrop()
                .into(profileImage)
        } else {
            profileImage.setImageResource(R.drawable.ic_account_circle_white)
        }

        val onlineRef = users.child(participant?.uid!!).child("online")

        onlineRef.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val online = snapshot.value.toString().toBoolean()

                if (online) {
                    statusOnline.visibility = View.VISIBLE
                    statusOnline.text = getString(R.string.conversation__status_online)
                } else {
                    statusOnline.visibility = View.GONE
                }
            }

            override fun onCancelled(error: DatabaseError) {
                Log.e(TAG, "Error reading 'online' value: ${error.message}")
            }
        })

        chatExists = ConversationUtil.conversationExists(user!!, participant!!)

        if (chatExists) {
            val existingConversationId =
                ConversationUtil.getConversationId(participant!!, user!!.conversations)

            if (!existingConversationId.isNullOrEmpty()) {
                currentConversationId = existingConversationId
                prepareConversation(existingConversationId)
            } else {
                Log.e(TAG, "Conversation ID not found despite chat existence")
            }
        }
    }

    private fun prepareConversation(conversationId: String) {
        val previousSecret = conversationSecret
        conversationSecret = null

        ConversationKeyManager.ensureConversationSecret(
            this,
            conversationId,
            user!!,
            participant!!,
            onSuccess = success@{ secret ->
                if (previousSecret == secret && ::conversationUtil.isInitialized) {
                    conversationSecret = secret
                    adapter?.setConversationSecret(secret)
                    return@success
                }

                conversationSecret = secret
                conversationUtil = ConversationUtil(this, conversationId, user!!, participant!!, secret)
                loadConversation(conversationId)
            },
            onError = { error ->
                Log.e(TAG, "Failed to establish secure channel", error)
                conversationSecret = previousSecret
                previousSecret?.let { adapter?.setConversationSecret(it) }
                Util.showMessage(this, "Unable to secure the conversation. Please try again.")
            }
        )
    }

    private fun loadConversation(cid: String) {
        val messages: Query = conversations.child(cid).child(Key.Conversation.MESSAGES)

        val options =
            FirebaseRecyclerOptions.Builder<Message>().setQuery(messages, Message::class.java)
                .build()

        adapter?.stopListening()

        adapter = MessageAdapter(applicationContext, options).apply {
            setConversationId(cid)
            setParticipant(participant)
            conversationSecret?.let { setConversationSecret(it) }

            registerAdapterDataObserver(object : AdapterDataObserver() {
                override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                    super.onItemRangeInserted(positionStart, itemCount)
                    recyclerview.smoothScrollToPosition(positionStart)
                }
            })
        }

        recyclerview.adapter = adapter
        adapter?.startListening()
    }

    override fun onStart() {
        adapter?.startListening()
        super.onStart()
    }

    override fun onStop() {
        adapter?.stopListening()
        super.onStop()
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onResume() {
        UserUtils.update("online", "true")
        adapter?.startListening()
        adapter?.notifyDataSetChanged()
        super.onResume()
    }
}
